### PR TITLE
Reintroduce ExitType

### PR DESCRIPTION
**Pull request by unusual-thoughts**
This PR introduces `ExitType=main|cgroup` for services. Similar to how `Type` specifies the launch of a service, `ExitType` is concerned with how systemd determines that a service exited. 

- If set to `main` (the current behavior), the service manager will consider the unit stopped when the main process exits.

- The `cgroup` exit type is meant for applications whose forking model is not known ahead of time and which might not have a specific main process. The service will stay running as long as at least one process in the cgroup is running. This is intended for transient or automatically generated services, such as graphical applications inside of a desktop environment.

Motivation for this is #16805 . `xdg-autostart-generator` was amended to use `ExitType=cgroup`, which fixes a [bug](https://bugs.kde.org/show_bug.cgi?id=433299) with some generated autostart app services that forked not launching (systemd killed the entire cgroup after the main process exited)

The original PR (#18782) was reverted (#20073) after realizing that the exit status of "the last process in the cgroup" can't reliably be known (#19385)

This version instead uses the main process exit status if there is one and just listens to the cgroup empty event otherwise.

### CODE REVIEW

File: src/core/service.c [hunk 1]
```diff
@@ -3385,7 +3387,13 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
...
                 assert_not_reached();
 
-        if (s->main_pid == pid) {
+        /* Services with ExitType=cgroup ignore the main PID for purposes of exit status */
+        if (s->exit_type == SERVICE_EXIT_CGROUP && s->main_pid == pid) {
+                service_unwatch_main_pid(s);
+                s->main_pid_known = false;
+        }
+
+        if ((s->exit_type == SERVICE_EXIT_MAIN && s->main_pid == pid)) {
```
Comment (poettering) [inline src/core/service.c]: one () is redundant

File: src/core/service.c [hunk 2]
```diff
@@ -3385,7 +3387,13 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
         else
                 assert_not_reached();
 
-        if (s->main_pid == pid) {
+        /* Services with ExitType=cgroup ignore the main PID for purposes of exit status */
+        if (s->exit_type == SERVICE_EXIT_CGROUP && s->main_pid == pid) {
+                service_unwatch_main_pid(s);
+                s->main_pid_known = false;
```
Comment (poettering) [inline src/core/service.c]: The exec_fd_event_source should be cleaned up still in this case, i.e the first part of the if block below should run in that case too

Comment (poettering) [reply src/core/service.c]: I think service_load_pid_file() and so on should also still run. i.e. we should still track the main pid as good as we can in this mode. Knowing what the main pid is is one thing, considering the service down whenever the main pid is gone is another.

**Comment by poettering** [comment]
As mentioned above, I think we should still track the main pid if we know it. i.e. this new option should turn off acting on it exiting, that's all.

**Comment by unusual-thoughts** [comment]
> As mentioned above, I think we should still track the main pid if we know it. i.e. this new option should turn off acting on it exiting, that's all.

I made the changes you suggested.
Now the main PID is still tracked if defined, and does end up being the service exit status when it eventually dies from the cgroup becoming empty. 
If no main pid is defined or no sigchld is received from it, the service exits successfully

**Comment by unusual-thoughts** [comment]
@bluca can you take a look ?

File: test/units/testsuite-56.sh [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+# process tree: systemd -> bash -> bash -> sleep
+((sleep infinity); true) &
+
+# process tree: systemd -> bash -> sleep
+sleep infinity
+EOF
+chmod +x /tmp/test56-exit-cgroup.sh
+
+# service should be stopped cleanly
+(sleep 1; systemctl stop one) &
```
Comment (bluca) [inline test/units/testsuite-56.sh]: on slow virtual machines this might be a bit racy, make it 10?

File: man/systemd.service.xml [hunk 4]
```diff
@@ -255,6 +255,30 @@
...
+      <varlistentry>
+        <term><varname>ExitType=</varname></term>
+
+        <listitem>
+          <para>Configures the process exit type for this service unit. One of <option>main</option> or
+          <option>cgroup</option>:</para>
+
+          <itemizedlist>
+            <listitem><para>If set to <option>main</option> (the default), the service manager
+            will consider the unit stopped when the main process, which is determined according to the `Type`, exits.
```
Comment (poettering) [inline man/systemd.service.xml]: `Type` → `<varname>Type=</varname>`

File: src/core/service.c [hunk 5]
```diff
@@ -3269,6 +3271,13 @@ static void service_notify_cgroup_empty_event(Unit *u) {
                         else
                                 service_enter_stop(s, SERVICE_FAILURE_PROTOCOL);
                 }
+
+                if (s->exit_type == SERVICE_EXIT_CGROUP) {
+                        if (s->state == SERVICE_START)
+                                service_enter_stop_post(s, SERVICE_SUCCESS);
```
Comment (poettering) [inline src/core/service.c]: this doesn't appear right to me. if you have a unit with:

```
ExecStart=/bin/true
ExecStartPost=/bin/true
```

Then you might nevre run the second /bin/true here because you immediately skip into ExecStopPost=. It appears to me you should never act on either SERVICE_START not SERVICE_START_POST

File: src/core/service.c [hunk 6]
```diff
@@ -3280,7 +3289,8 @@ static void service_notify_cgroup_empty_event(Unit *u) {
         case SERVICE_STOP_SIGTERM:
         case SERVICE_STOP_SIGKILL:
 
-                if (main_pid_good(s) <= 0 && control_pid_good(s) <= 0)
+                if (s->exit_type == SERVICE_EXIT_CGROUP ||
```
Comment (poettering) [inline src/core/service.c]: this appears unnecessary a change no? if we know a main or control pid it's worth waiting for them

File: src/core/service.c [hunk 7]
```diff
@@ -3289,7 +3299,8 @@ static void service_notify_cgroup_empty_event(Unit *u) {
         case SERVICE_FINAL_WATCHDOG:
         case SERVICE_FINAL_SIGTERM:
         case SERVICE_FINAL_SIGKILL:
-                if (main_pid_good(s) <= 0 && control_pid_good(s) <= 0)
+                if (s->exit_type == SERVICE_EXIT_CGROUP ||
+                    (main_pid_good(s) <= 0 && control_pid_good(s) <= 0))
```
Comment (poettering) [inline src/core/service.c]: as above

File: src/core/service.c [hunk 8]
```diff
@@ -3428,9 +3439,6 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
                                 f == SERVICE_SUCCESS,
                                 code, status);
 
-                if (s->result == SERVICE_SUCCESS)
-                        s->result = f;
```
Comment (poettering) [inline src/core/service.c]: why move this? I think if we know the main process we should still consider its exit status more relevant than the others and track it.

File: src/core/service.c [hunk 9]
```diff
@@ -3441,7 +3449,10 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
...
                         service_run_next_main(s);
 
-                } else {
+                /* Services with ExitType=cgroup do not act on main PID exiting */
+                } else if (s->exit_type == SERVICE_EXIT_MAIN) {
+                        if (s->result == SERVICE_SUCCESS)
+                                s->result = f;
 
                         /* The service exited, so the service is officially gone. */
                         s->main_command = NULL;
```
Comment (poettering) [inline src/core/service.c:3486]: i am pretty sure this invalidation should still run even in cgroup exit mode

File: src/core/service.c [hunk 10]
```diff
@@ -3243,7 +3243,9 @@ static void service_notify_cgroup_empty_event(Unit *u) {
                  * because it includes return codes/signals. Which is
                  * why we ignore the cgroup events for most cases,
                  * except when we don't know pid which to expect the
-                 * SIGCHLD for. */
+                 * SIGCHLD for. In the case of ExitType=cgroup,
+                 * we don't based act on the main PID exiting and
+                 * instead we do listen to the cgroup event */
```
Comment (behrmann) [inline src/core/service.c]:
```suggestion
                 * we don't act on the main PID exiting and
                 * instead we listen to the cgroup event */
```

File: src/core/service.c (hunk 8 above)
Comment (unusual-thoughts) [reply src/core/service.c]: I thought delaying setting the result until the cgroup became empty made more sense, but I think you're right.

File: src/core/service.c (hunk 9 above)
Comment (unusual-thoughts) [reply src/core/service.c:3486]: That's true

File: src/core/service.c (hunk 6 above)
Comment (unusual-thoughts) [reply src/core/service.c]: True. However if we do wait for it (when the cgroup empty event comes just before the signal) we need to handle the state changes when the signal does happen.
I guess that can be done with a check for either `SERVICE_EXIT_MAIN` or the cgroup being empty before the main switch on `s->state` in `service_sigchld_event()`

File: test/units/testsuite-56.sh (hunk 3 above)
Comment (unusual-thoughts) [reply test/units/testsuite-56.sh]: 10 sounds a bit excessive ? there are 5 of these

File: test/units/testsuite-56.sh [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+# process tree: systemd -> bash -> bash -> sleep
+((sleep infinity); true) &
+
+# process tree: systemd -> bash -> sleep
+sleep infinity
+EOF
+chmod +x /tmp/test56-exit-cgroup.sh
+
+# service should be stopped cleanly
+(sleep 5; systemctl stop one) &
```
Comment (bluca) [inline test/units/testsuite-56.sh]: Builds with coverage and without optimization, on nested VMs without kvm on overworked nodes, will likely be racy. Can you make this deterministic? ie, use Type=notify or check status in a loop rather than an arbitrary sleep

Comment (unusual-thoughts) [reply test/units/testsuite-56.sh]: The sleep is only to delay stopping the unit after the /tmp/test56-exit-cgroup.sh bash script is executed to the end (which should take a few ms at most). 

The service itself is already being waited on by systemd-run in a deterministic way. I think this should be fine ?

Comment (bluca) [reply test/units/testsuite-56.sh]: The point is that there's no guarantee that it will work as you want it with just a sleep. It might take much longer than 5s to schedule the job and get to the point in the script where you want it to happen.

Comment (unusual-thoughts) [reply test/units/testsuite-56.sh]: I rewrote the tests to use `Type=nofify` and `ExecStartPost=` to have systemd itself shut down/kill the services precisely when needed

Comment (bluca) [reply test/units/testsuite-56.sh]: nice, much better

File: man/systemd.service.xml [hunk 12]
```diff
@@ -255,6 +255,31 @@
...
+
+        <listitem>
+          <para>Configures the process exit type for this service unit. One of <option>main</option> or
+          <option>cgroup</option>:</para>
+
+          <itemizedlist>
+            <listitem><para>If set to <option>main</option> (the default), the service manager
+            will consider the unit stopped when the main process, which is determined according to the
+            <varname>Type=</varname>, exits. Consequently, it cannot be used with
+            <varname>Type=</varname><option>oneshot</option></para></listitem>
```
Comment (behrmann) [inline man/systemd.service.xml]: Are you missing a a period here?
```suggestion
            <varname>Type=</varname><option>oneshot</option>.</para></listitem>
```

Comment (unusual-thoughts) [reply man/systemd.service.xml]: thanks

File: src/core/service.c [hunk 13]
```diff
@@ -3479,79 +3486,82 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
                         service_run_next_main(s);
```
Comment (unusual-thoughts) [inline src/core/service.c:3483]: ignoring whitespace, this diff looks like (`git show -b  src/core/service.c`):
```diff
@@ -3479,10 +3486,12 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
                         service_run_next_main(s);
 
                 } else {
-
-                        /* The service exited, so the service is officially gone. */
                         s->main_command = NULL;
 
+                        /* Services with ExitType=cgroup do not act on main PID exiting,
+                         * unless the cgroup is already empty */
+                        if (s->exit_type == SERVICE_EXIT_MAIN || cgroup_good(s) <= 0) {
+                                /* The service exited, so the service is officially gone. */
                                 switch (s->state) {
 
                                 case SERVICE_START_POST:
@@ -3554,6 +3563,7 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
                                         assert_not_reached();
                                 }
+                        }
                 }
 
         } else if (s->control_pid == pid) {
                 const char *kind;
```

File: man/systemd.service.xml [hunk 14]
```diff
@@ -255,6 +255,31 @@
         </listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>ExitType=</varname></term>
+
+        <listitem>
+          <para>Configures the process exit type for this service unit. One of <option>main</option> or
```
Comment (keszybz) [inline man/systemd.service.xml]: Maybe "Specifies when the manager should consider the service to be finished" ("configures the process exit type" doesn't really say much). You can also drop "One of mian or cgroup", the bullet list below already contains this.

File: man/systemd.service.xml [hunk 15]
```diff
@@ -255,6 +255,31 @@
...
+          <option>cgroup</option>:</para>
+
+          <itemizedlist>
+            <listitem><para>If set to <option>main</option> (the default), the service manager
+            will consider the unit stopped when the main process, which is determined according to the
+            <varname>Type=</varname>, exits. Consequently, it cannot be used with
+            <varname>Type=</varname><option>oneshot</option>.</para></listitem>
+
+            <listitem><para>The <option>cgroup</option> exit type is meant for applications whose forking model is not
+            known ahead of time and which might not have a specific main process. The service will stay running as long
```
Comment (keszybz) [inline man/systemd.service.xml]: Please drop the explanatory part of the first sentence here. Reference first, explanations laters.
Also maybe "stay running" → "be considered running".

`When  <option>cgroup</option>, the service will be considered running as long as at least one process in the cgroup has not exited.` ?

File: src/xdg-autostart-generator/xdg-autostart-service.c [hunk 16]
```diff
@@ -597,6 +597,7 @@ int xdg_autostart_service_generate_unit(
         fprintf(f,
                 "\n[Service]\n"
                 "Type=exec\n"
+                "ExitType=cgroup\n"
```
Comment (keszybz) [inline src/xdg-autostart-generator/xdg-autostart-service.c:600]: It'd be nice to split this chunk to a separate commit… It's a separate logical change and it'll makes bisecting easier to have two commits.

File: src/core/service.c [hunk 17]
```diff
@@ -3291,6 +3294,10 @@ static void service_notify_cgroup_empty_event(Unit *u) {
                         break;
                 }
 
+                if (s->exit_type == SERVICE_EXIT_CGROUP && main_pid_good(s) <= 0) {
+                        service_enter_start_post(s);
+                }
```
Comment (keszybz) [inline src/core/service.c]: Please no {} around one-line blocks.

**Review by keszybz — CHANGES_REQUESTED** [review]
Sorry for the delay. Looks very nice. I like that in the end the code change in src/core/service.c is minimal.

(As usual) I have some requests for the man page texts…

File: src/xdg-autostart-generator/xdg-autostart-service.c (hunk 16 above)
Comment (unusual-thoughts) [reply src/xdg-autostart-generator/xdg-autostart-service.c:600]: sure, will do

File: man/systemd.service.xml (hunk 14 above)
Comment (unusual-thoughts) [reply man/systemd.service.xml]: makes sense

File: man/systemd.service.xml (hunk 15 above)
Comment (unusual-thoughts) [reply man/systemd.service.xml]: Right. I'll merge the explanation with the rest below then.

File: man/systemd.service.xml (hunk 14 above)
Comment (unusual-thoughts) [reply man/systemd.service.xml]: Regarding dropping the enum values in the sentence, I followed the style of other parts of the docs, like with `Type=` or `TimeoutStartFailureMode`

**Comment by keszybz** [comment]
focal-amd64: TEST-50-DISSECT:                    FAIL     (316 s)
focal-arm64: logind               FAIL non-zero exit status 1
unit tests / gcc:  322/1232 test-repart                             TIMEOUT 30.01 s 

The usual ;]

File: man/systemd.service.xml [hunk 18]
```diff
@@ -268,15 +268,15 @@
...
+            least one process in the cgroup has not exited.</para></listitem>
           </itemizedlist>
 
           <para>It is generally recommended to use <varname>ExitType=</varname><option>main</option> when a service has
           a known forking model and a main process can reliably be determined. <varname>ExitType=</varname>
-          <option>cgroup</option> is well suited for transient or automatically generated services, such as graphical
-          applications inside of a desktop environment.</para>
+          <option>cgroup</option> is meant for applications whose forking model is not known ahead of time and which
+          might not have a specific main process. It is well suited for transient or automatically generated services,
+          such as graphical applications inside of a desktop environment.</para>
```
Comment (keszybz) [inline man/systemd.service.xml]: This whole chunk belongs in the previous commit. I'll force push to move it.

Comment (keszybz) [reply man/systemd.service.xml]: Hmm, I can't push to your branch. Not sure why. Please move this hunk to the previous patch.

Comment (unusual-thoughts) [reply man/systemd.service.xml]: Oops, should be fixed now

**Comment by poettering** [comment]
Love how minimal this eventually became! Good work!
